### PR TITLE
feat: rbac middleware

### PR DESCRIPTION
**Pull request by zlwaterfield**
## Changes

Adding middleware for RBAC. This is mostly pulling the logic from @benjackwhite  previous PR: #20864 but updating a bunch more tests and a few logic pieces. 

See https://github.com/PostHog/posthog/issues/24512 for more information.

Key additions:
- ee
  - AccessControlSerializer - for the AccessControl model
  - AccessControlViewSetMixin - Adds an "access\_controls" action to the viewset that handles access control for the given resource
- posthog
	- AccessControlPermission - Unified permissions access - controls access to any object based on the user's access controls
	- UserAccessControlSerializerMixin - Mixin for serializers to add user access control fields
	- UserAccessControl - UserAccessControl provides functions for checking unified access to all resources and objects from a Project level downwards.

Many of the changes are test updates for the number of queries. Then we've also added the middleware to the viewsets of Dashboard, Projects, Organizations, Insights, Notebooks, Feature Flags, etc.

More details in Readme to come.

👉 _Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review._

## How did you test this code?

Updated existing tests
Added new tests

### CODE REVIEW

File: posthog/api/routing.py [hunk 1]
```diff
@@ -145,19 +147,47 @@ def dangerously_get_queryset(self) -> QuerySet:
...
+        if getattr(self, "_in_get_queryset", False):
+            return super().get_queryset()
 
-        queryset = super().get_queryset()
-        # First of all make sure we do the custom filters before applying our own
         try:
-            queryset = self.safely_get_queryset(queryset)
-        except NotImplementedError:
-            pass
+            self._in_get_queryset = True
```
Comment (benjackwhite) [inline posthog/api/routing.py:155]: Ooh this is nice

**Review by benjackwhite — APPROVED** [review]
This looks solid! (If i do say so myself).

Should be totally safe but we might want to be ready for a qucik revert just in case

File: posthog/api/organization.py [hunk 2]
```diff
@@ -127,7 +129,14 @@ def get_membership_level(self, organization: Organization) -> Optional[Organizat
         return membership.level if membership is not None else None
 
     def get_teams(self, instance: Organization) -> list[dict[str, Any]]:
-        visible_teams = instance.teams.filter(id__in=self.user_permissions.team_ids_visible_for_user)
+        # Support new access control system
```
Comment (zlwaterfield) [inline posthog/api/organization.py:132]: Review this logic

File: posthog/permissions.py [hunk 3]
```diff
@@ -19,13 +19,38 @@
...
+            return object.team.organization  # type: ignore
+        except AttributeError:
+            try:
+                return object.project.organization  # type: ignore
+            except AttributeError:
+                pass
+    raise ValueError("Object not compatible with organization-based permissions!")
+
+
+def extract_organization_membership(object: Model, view: ViewSet) -> Organization:
```
Comment (zlwaterfield) [inline posthog/permissions.py:53]: Remove dup function

**Pull request merged**